docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,137 @@
+
+## Summary
+
+The new type solver for Luau supports a limited number of built-in type functions that enable developers to be expressive with their code. To further enhance this expressivity and the flexibility of the type system, this RFC proposes the design and implementation of user-defined type functions, a feature that will allow developers to define custom type functions.
+
+## Motivation
+
+The primary motivation for introducing user-defined type functions is to extend the flexibility of the type system and empower developers to create more expressive and type-safe libraries. Current limitations in the type system restrict custom type manipulations, which are crucial for creating advanced patterns and abstractions. By supporting user-defined type functions, we aim to:
+- Enable more precise type definitions that can capture complex relationships and constraints
+- Facilitate the creation of reusable and composable libraries
+- Enhance code safety and maintainability with better type-level programming
+
+The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, ultimately improving the Luau developer experience.
+
+## Design
+
+Type functions wil be defined with the following syntax:
+```luau
+type function f(...)
+    -- implementation of the type function
+end
+```
+
+For instance, the `rawget` type function can be written as:
+```luau
+type function rawget(tbl, prop)
+    return tbl.getProps()[prop.getValue()]
+
+    -- returning nothing / nil means that this type function will fail to reduce
+end
+
+type Person = {
+    name: string,
+    age: number
+}
+
+type ty = rawget<Person, "name"> -- ty = string
+```
+To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `lType`.  An `lType` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, each argument of a type function is serialized into a userdata called `lType`. Moreover, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 
+
+<details><summary>lType library methods (dropdown)</summary>
+Note: methods under a different type heading (ex: `Singleton`) imply that the methods are only available for those types. At the implementation level, there is a check to make sure that the type-specific methods are being called on the correct types (e.g, for `getIndexer()`, assert that `isTable()` is true).
+
+#### Any
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `isstring()` | `boolean` | returns true if self is of type `string` |
+| `isnumber()` | `boolean` | returns true if self is of type `number` |
+| `isboolean()` | `boolean` | returns true if self is of type `boolean` (e.g. true or false) |
```

Comment:
Sorry, I didn't mean to make a note that `boolean` is equivalent to `true | false`. What I meant to say was there's both `isboolean()` and `isbooleansingleton()` below, so the former felt redundant.

I guess what I'm really asking here is, when a type argument is marshalled into a type function, what are we _really_ passing in? e.g.

```lua
type T = { p: string } & { q: number }
type U = { p: string, q: number }
```
Is `my_type_fun<T>` passed the same structure as `my_type_fun<U>`?

Reply:
I would imagine they would get different types because type functions would work syntactically, but an alternative approach would be using normalization to canonize the type. Unfortunately, normalization doesn't actually normalize tables very much as-is though, so that would require a bunch of additional work on our part to expand the scope of normalization. One possibility here is that we take it syntactically in whatever form it is, and we could consider adding normalization (or some of the normalization APIs, like `isIntersectionInhabited`) to the API surface for the types library, but that is a bit scary.

Reply:
I kind of want to avoid making promises on this front at least until egraphs is fully implemented, then we can always pass in a type that's guaranteed to be most simplest, for some definition of simplest.

Then users won't have to worry about dealing with cases like `string & ~"a"` vs `{ p: number } & { q: string }`.